### ISSUE TITLE
PTT/keybindings to F13-F24 

### ISSUE THREAD

**Issue by mumble-voip** [mumble-voip/mumble#987]
In 1.2.4-rc1-13-gc00b44e mapping key shortcuts to F13-F24 does not work. I found another (now closed) issue with exactly the same problem but it claimed to have been fixed in beta 49 of 1.2.4.

_This ticket has been migrated from sourceforge. It is thus missing some details like original creator etc.
The original is at https://sourceforge.net/p/mumble/bugs/987/ ._

**Comment by GuiSim** [comment]
I was able to reproduce this issue in 1.2.8.

**Comment by Volatar** [comment]
This issue persists currently as of the 1.3.0~687~g9c462ce~snapshot.

**Comment by forivall** [comment]
@SuperFlue as far as the g710+ goes, #1730 should be able to take care of that problem.

**Comment by Volatar** [comment]
Not everyone has access to the G keys. Everyone can bind F13-24 though. (Heck, you can find keyboards with those keys if you look hard enough.)

**Comment by forivall** [comment]
@Volatar I'm not claiming to solve this problem at all :cry: just enough for logitech users to get by :angry:

**Comment by mkrautz** [comment]
@Volatar 

Which program do you wish Mumble to receive these keys from? Autohotkey?

Having a reference in here should allow anyone picking this bug up to have a way to reproduce the issue.

**Comment by pyrojoe** [comment]
Mumble receiving keybinds from Autohotkey would be ideal because AHK opens up many options for macros.  Mumble DOES acknowledge ahk binds if ahk is run as admin but mapping F13 won't map.  Using the exact same script will map F12 so I'm not sure what the issue is.  I used to be able to map F13 on Mumble v1.2.3 and I had a ticket found [here.](http://sourceforge.net/p/mumble/bugs/941/) where I mentioned it.  I recall someone saying that the keybind code didn't change between 1.2.3 and 1.2.4 but something changed there, possibly a dependency? I'm not really familiar with C++ or the codebase so wouldn't know.  Upgrading to 1.2.4 from 1.2.3 keeps previously entered F13 keybinds (which is what my ticket was about, a snapshot version broke the bind) but the bind just shows as "K:" not "K:F13" kinda buggy.

My current script is just running:
```
#NoEnv
#InstallKeybdHook
SendMode Input
SendInput {F13}  
ExitApp
```

Something else that just occurred to me that might be related.  The F13 bind never worked while Mumble is in focus I don't know why, but that could be the reason it won't bind.. because it doesn't work with mumble in focus.

**Comment by forivall** [comment]
Perfect! When I have the time, using that information, I'll be able to perform a `git bisect` to figure out exactly which commit broke it. Personally, looking at the difference between 1.2.3 and 1.2.4, it could be a number of different changes that had this unexpected side-effect.

**Comment by pyrojoe** [comment]
Before you go through the trouble let me roll back my Mumble to 1.2.3 and verify that autohotkey works for setting the bind.  This was 2 years ago and I recall using an old version of Logitech's setpoint software to bind F13 to my G500 mouse button.  I can't remember if I had to set the bind on Mumble with ahk or my mouse.  Not sure if that little detail matters or not since I do know that binding at all was doable in 1.2.3 but not 1.2.4.  I'll update once I figure out how I did it.

**Comment by forivall** [comment]
Yeah, I use [this script](https://gist.github.com/forivall/3e223b27ff19c2fe4009) in the logitech gaming software (LGS), and when I was doing some research for the native g-key support (and research for a teamspeak tool+plugin), I did discover that the f13-18 generated by LGS were being recognized by the [nil library](https://github.com/noorus/nil).

**Comment by pyrojoe** [comment]
Ok so I was completely wrong about 1.2.3 working but anything after that not working.. my bad.  Getting it to work required a bit of tinkering and I guess I never figured out the exact setup that worked.  You can map F13 on the current version of mumble the same way you can map it on 1.2.3 and it can be done with just ahk.  In order to set the hotkey I needed two scripts running as **admin**.  
In this script I bound a key on my keyboard to F13.
Using the below script by itself doesn't really work, if you try to set the shortcut by pressing 'o' mumble just sits there waiting for a button push as if you haven't clicked anything.  If F13 is already mapped then 'o' will act as a shortcut.
```
#NoEnv
#InstallKeybdHook
SendMode Input
o::F13
```
Now.. if you keep the first script open and running and use this script which sends an F13 keypress on a 4 second delay (so you have time to start the script then click the shortcut area in mumble) then you can set F13 as a short cut.
```
#NoEnv
#InstallKeybdHook
SendMode Input
sleep, 4000
SendInput {F13}
ExitApp
```
This second script isn't needed if you already have a macro button or whatever mapped to F13, you can just use that button instead of the second script.

**Comment by Volatar** [comment]
I have not been using AHK. I was using the little Java app that @SuperFlue mentioned earlier to bind extra buttons to F13 and friends. Unless I am mistaken, the results should be a native windows keypress of F13, nothing crazy, nothing scripted, just a normal keypress. And said keypress does not work on the current version of mumble.

**Comment by Volatar** [comment]
A quick test of using the Java app to send F13 to mumble's key configuration instead of pressing my macro did not result in success by the way.

**Comment by jkluch** [comment]
@Volatar Are you having trouble mapping F13 in mumble or are you having trouble using an existing F13 shortcut in mumble?  If you're unable to map the key, try the method below:

Do me a favor and install ahk.  create a file named something.ahk with the contents
```
#NoEnv
#InstallKeybdHook
SendMode Input
o::F13
```

This will map the 'o' key (just a random key) to F13.  After you run that script try to use your Java app to add the shortcut.  If it works you can then stop the script by closing the program in your notification tray and you should be able to use your Java app without needing the ahk script.

If that doesn't work create the second script like I mentioned in the post above yours and try that.

I'm trying to figure out if there's some logic being done to check for an F13 key before allowing the bind and if by using the AHK script you're tricking mumble into thinking there is an F13 key.

**Comment by Volatar** [comment]
@jkluch "Are you having trouble mapping F13 in mumble or are you having trouble using an existing F13 shortcut in mumble?"

Both

I shall try your directions momentarily and return and report.

**Comment by Volatar** [comment]
Running the first script and then attempting to configure mumble using the Java app did not work. (Though the script was doing something. Had to stop it to type the letter o here.)

Note, your statement here is false for me: "Using [the first script] by itself doesn't really work, if you try to set the shortcut by pressing 'o' mumble just sits there waiting for a button push as if you haven't clicked anything."

What happened when I tried that was that mumble accepted the keypress as the normal press of the 'o' key. This push to talk key configuration did NOT work when I then tried to use it. It simply would scroll through anything starting with 'o' on the server and did not execute push to talk, even with Mumble not focused.

For testing, I opened up Teamspeak 3, which I already have configured to use F13 as push to talk, and with that first script running it did not register 'o' as F13, but also took it as the letter, executing its standard behavior of selecting the first thing in the server starting with 'o', much like Mumble.

I then tried configuring TS3 with the script running and, like Mumble, it took the keypress as the letter 'o' (while an actual input of F13 to TS3 shows up in the configuration screen as '124'). Unlike mumble, this configuration of 'o' as PTT (AHK script still running) DID work as push to talk.

Trying second script and your directions now.

Second script did not behave as instructed. Mumble acted as though no key had been pressed at all, just as it does with the Java app and my configured macro. Teamspeak blipped though. It got the F13 press, so the script works. I then closed TS3 to make sure it was not interfering in any way and got the same results with Mumble.

I suspect it might be possible that your first script is doing something wrong? The fact that when it is running, neither Mumble nor TS3 recognize pressing 'o' as F13, yet Chrome does (as I cannot type the letter 'o' with the script running), seems.... weird.

**Comment by jkluch** [comment]
Just to be sure did you launch these scripts as admin?  It doesn't sound like it.  You need to right click the file, then press the "run as administrator" option like so: http://i.imgur.com/ZHhTb1y.png

**Comment by Volatar** [comment]
I did not use that option. Let me try that...

Ah ha! Script #1 had no different behavior and did not allow me to use the Java app to send the key to mumble, but upon utilizing the two scripts together (run in admin mode), Mumble detected the F13 press. The shortcut shows up blank, but both 'o' while script 1 is running, and my mouse macro to F13 now work as push to talk in Mumble.

I think we are getting somewhere! :)

**Comment by forivall** [comment]
If you run the java app as admin, does that work?

**Comment by Volatar** [comment]
I do not know how to run a .jar as administrator.

(Windows 8.1)

**Comment by forivall** [comment]
if you know how to run `cmd` as administrator in 8.1 (I use 7), do that and then try `java -jar <path-to-jarfile>`. Hopefully java is in your `%PATH%`. Or else you'd have to do `<path-to-java> -jar <path-to-jarfile>`. :sob:

**Comment by Volatar** [comment]
Opening cmd in admin mode is as easy as right clicking the start button in 8.1. One of its redeeming features for a power user I guess.

Tested. No change. Did not work. Hmmm...

**Comment by andrewpmartinez** [comment]
This works in Mumble 1.3.0~867~g197f13e~snapshot 64bit, on Windows 10 64bit.

The key binding will be displayed as "K: " instead of "K: F13".

**Comment by Etuldan** [comment]
Don't know why, but suddenly (after computer restart), mumble lost my PTT setting to F13, and I'm not able to set it again (keybinding not recognized), nor with an AHK script, nor with my already configured Logitech mouse.
I tried also to upgrade my mumble installation (from 1.2.12 to 1.2.13, both stable) without success.
Windows 7 x64

**Comment by forivall** [comment]
I'm mostly curious, but could anyone who's still having this problem see if it's detected by https://github.com/forivall/niftykb/releases? <- program where I pulled out mumble's keybinding code into a standalone app. This way, we'll know if it's a problem with the keybinding code itself, or something else in mumble is interfering with the keybinding.

**Comment by Etuldan** [comment]
@forivall "This application failed to start because it could not find or load the Qt platform plugin "windows" ".
I unzip all the content (including the QT5 related DLLs) from https://github.com/forivall/niftykb/releases/download/v0.1.0/niftykb.zip (even with QT 5.5.1 installed)

**Comment by Volatar** [comment]
This issue persists on 1.3.0~970~g888a459~snapshot for win-64 on my Windows 8.1 64 bit machine.

I was unable to make your program work @forivall.

**Comment by Glockshna** [comment]
I was able to bind F13 through my Logitech Gaming Software to mumble on windows 7 but since I switched to Windows 10 Mumble no longer recognizes the key. It still works fine in Teamspeak though. Really getting frustrated with this =/

**Comment by Etuldan** [comment]
Found something new that may be interesting:

I need to run my autohotkey script which redirect a random key (F1) in my case to F13, in order to mumble accept any F13 key (simulated by my autohotkey script of by my Logitech mouse)
Without this script, even if I press on my F13 button on my mouse, I can't configure mumble nor use the previously PTT configured function.
All is OK w/o the autohotkey script running for teamspeak for example (they have both the same key bind)

So, here is my conclusion:
By default, as my keyboard don't have this F13 key, mumble won't accept it from any device, even if it works for other software. As soon as I add this key (with the autohotkey script), mumble will accept this key.

I'm therefore not 100% sure about this, I'll continue to inspect on my Win7 64b installation.

**Comment by mkrautz** [comment]
I can reproduce this when Mumble is run in with uiAccess=true (making Mumble an UI Automation client).
Mumble is run in this mode when installed into `Program Files`, or `Program Files (x86)`.

But if I copy my Mumble installation to my Desktop, and run it from there, it works fine.
(Running Mumble outside of `Program Files` causes UI Automation to be disabled).

I tried binding F1 to F13, using the script provided by @Etuldan.

When running Mumble like this, the shortcut name in the shortcut settings is simply `K:`, because my keyboard doesn't have an F13 key.

But the shortcut itself works just fine.

Can anyone subscribed to this issue see if this workaround works for them?
1. Copy the Mumble directory from `c:\Program Files\Mumble` or `c:\Program Files (x86)\Mumble` to your desktop.
2. Run Mumble from your desktop.
3. Try using your F13-F24 key in Mumble.

Thanks.

**Comment by Keridos** [comment]
Doesnt work for me in mumble-1.3.0~1797~g6ac0553~snapshot.winx64 on windows 10 64-bit. Even with the client being run from my desktop it does not work.

**Comment by zkxs** [comment]
Using Mumble 1.2.18 on Windows 7 from the MSI, I can successfully use F13 as PTT.  This is with Mumble installed in Program Files.  

I used [This .reg file](https://gist.github.com/zkxs/2d570350489596145d956eeb55fe0562) to rebind Caps Lock to F13.  Worth noting is that even though the shortcut bind works as expected, it is displayed as `K:Caps Lock`

**Comment by Keridos** [comment]
I have additional info. In all versions I tested it sporadically works, but after a pc restart I have to reset the keybinding and I do need to start autohotkey and use that to map another key to f13. After i used that I can use my mouse button that works as F13 as intended.

**Comment by crimsonblod** [comment]
I am also having this issue in all versions shown [here](http://i.imgur.com/X8XDYrB.png). And binding one of the keys on my keyboard to trick mumble into accepting f13 as an input doesn't work either.

Just figured I'd mention that this is still an issue, because it's been a few months since the last update here.

**Comment by jkluch** [comment]
Have you tried what I mention above?
I think you only need one script but according to what I wrote above I used two so try two.  
Using ahk create two files, like these, they bind 'o' and 'p' to F13

**f13_o.ahk**
```
#NoEnv
#InstallKeybdHook
SendMode Input

o::F13
```
**f13_p.ahk**
```
#NoEnv
#InstallKeybdHook
SendMode Input

p::F13
```

- If you have [autohotkey](https://autohotkey.com/download/) installed you should be able to right click your scripts and run as administrator, if you run the scripts with lower privileges the bind won't be recognized in mumble.
- Once both scripts are running open mumble's shortcut settings and  in the shortcut box press 'o' or 'p'  you should see "K:" without anything after it.  This is a bug in mumble but it means the bind worked.
- Mumble has another bug where the F13 bind doesn't work while the mumble window is in focus so click off of the mumble window before testing.

**Comment by darkuranium** [comment]
@mkrautz: Tried it. Doesn't work from the desktop, either.

FWIW, TeamSpeak and Discord are both installed in Program Files, and F13 works fine there. I don't use Ventrilo, but apparently, F13 works there, too.
Mumble is the ***only*** major VOIP software where it does not work, so I doubt the problem is anywhere *but* in Mumble.

**Comment by mkrautz** [comment]
Can people please retry this with the latest snapshot?

WinX64: https://dl.mumble.info/mumble-1.3.0~2569~gd196a4b~snapshot.winx64.msi

We just fixed @darkuranium's problem with F13.

If it doesn't work at first, you can also play with the option "Enable shortcuts in privileged applications" in Settings -> Shortcuts -> Additional Shorcut Engines.

Thanks!

**Comment by SuperFlue** [comment]
Since i am still subscribed to this....

F13-F24 works on the newest snapshot, tested using an Ardunio chip with the HID Project library to send a real USB HID key code.

Key binding shows up as "k:" but is otherwise working.

**Comment by PaletzTheWise** [comment]
Tested on https://dl.mumble.info/mumble-1.3.0~2586~g894ade2~snapshot.winx64.msi and https://dl.mumble.info/mumble-1.3.0~2586~g894ade2~snapshot.msi.
Both works with F16 generated by AHK.
I also tested with my mouse using Logitech gaming software where I bound F12 to mouse 4 and then edited the profile xml file to change it to F16. This works as intended with Teamspeak and x86 Mumble. However x64 Mumble detected mouse 4. In fact the mouse button was capable of clicking on Mumble UI elements.

**Comment by crimsonblod** [comment]
I'm still on 1.3.0~2569~gd196a4b and it's working for me on this build. If the link works the same way, this in theory should be the link to it. No promises though. If it's the right one, you should try it and see if your mouse works with it.
https://dl.mumble.info/mumble-1.3.0~2569~gd196a4b~snapshot.winx64.msi

**Comment by PaletzTheWise** [comment]
After a bit more testing looks like there is no difference in x64 and x86 since I have observed both versions become aware of physical mouse buttons. When it happens Mumble behavior depends on whether it is in focus or in background. It detects physical mouse buttons when focused, ignoring the F16 configured in Logitech gaming software. When in background it is the other way around, it detects the F16 and does not detect physical mouse button being pressed. This means I can use the AHK to configure a short-cut for F16 and then use the mouse-generated F16 once Mumble is in background.

I have not seen this physical button awareness in Mumble 1.2.4. The interesting part is that two days ago the x64 version was always aware, today the x86 version is always aware. This persists through Mumble restarts. Maybe it's down to which version I ran first. Finally, this awareness is not tied into F13-F24 since it affects other keys as well.
